core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java

```diff
@@ -1478,10 +1488,32 @@ public ServerBuilder contentPreview(int length) {
         return contentPreview(length, ArmeriaHttpUtil.HTTP_DEFAULT_CONTENT_CHARSET);
     }
 
+    /**
+     * Sets the {@link RequestConverterFunction}s, {@link ResponseConverterFunction}
+     * and {@link ExceptionHandlerFunction}s for creating a {@link AnnotatedHttpServiceExtensions}.
+     *
+     * @param requestConverterFunctions the {@link RequestConverterFunction}s
+     * @param responseConverterFunctions the {@link ResponseConverterFunction}s
+     * @param exceptionHandlerFunctions the {@link ExceptionHandlerFunction}s
+     */
+    public ServerBuilder annotatedHttpServiceExtensions(
+            Iterable<? extends RequestConverterFunction> requestConverterFunctions,
+            Iterable<? extends ResponseConverterFunction> responseConverterFunctions,
+            Iterable<? extends ExceptionHandlerFunction> exceptionHandlerFunctions) {
+        virtualHostTemplate.annotatedHttpServiceExtensions(requestConverterFunctions,
+                                                           responseConverterFunctions,
+                                                           exceptionHandlerFunctions);
+        return this;
+    }
+
     /**
      * Returns a newly-created {@link Server} based on the configuration properties set so far.
      */
     public Server build() {
+        final AnnotatedHttpServiceExtensions extensions =
+                virtualHostTemplate.getAnnotatedHttpServiceExtensions();
+        annotatedServiceBindingBuilders.forEach(builder -> builder.applyToServiceConfigBuilder(extensions));
```

Comment:
nit: need `assert extensions != null;`

Reply:
I think `virtualHostTemplate.getAnnotatedHttpServiceExtensions()` could be nullable. 🤔
https://github.com/line/armeria/blob/0b3e3fee81d7268922cb4a442e6d2571b55e1f6d/core/src/main/java/com/linecorp/armeria/server/VirtualHostBuilder.java#L154-L155

Reply:
Oh, I don't remember why I put it over there 😅

I'll fix it soon.

Reply:
This is somewhat unclear to me how it works because the `annotatedServiceBindingBuider` has the reference of the `serverBuilder` and call `addServiceConfigBuilder` inside in it.
How about just returning the list of `ServiceConfigBuilder` and add it here?
```java
List<ServiceConfigBuilder> toServiceConfigBuilders(AnnotatedHttpServiceExtensions extensions) {
    ...
    return elements.stream()
                   .map(element -> {
                       final HttpService decoratedService =
                               element.buildSafeDecoratedService(defaultServiceConfigSetters.decorator());
                       return defaultServiceConfigSetters.toServiceConfigBuilder(element.route(), decoratedService);
                   }).collect(toImmutableList());
}

...
// In ServerBuilder.build()
annotatedServiceBindingBuilders.forEach(annotatedBuilder -> {
    annotatedBuilder.toServiceConfigBuilders(extensions).forEach(
            virtualHostTemplate::addServiceConfigBuilder);
});
```